Update for NetBox 4.0.7; Add more logging, error checking; Fix #134, …

NicJames2378 opened the pull request:
Verified working with:

- NetBox 4.0.7
  - [hosted in Docker](https://github.com/linuxserver/docker-netbox)
  - Running on CentOS Stream 8
- Python 3.12.4
  - Running under Windows 11 22H2

NicJames2378 committed:
Update for NetBox 4.0.7; Add more logging, error checking; Fix #134, #84 (update)

brandan-schmitz commented on settings.py:
```diff
@@ -12,6 +12,13 @@
 NETBOX_TOKEN = os.getenv("NETBOX_TOKEN")
 IGNORE_SSL_ERRORS = (os.getenv("IGNORE_SSL_ERRORS", default="False") == "True")
 REPO_PATH = f"{os.path.dirname(os.path.realpath(__file__))}/repo"
+RETRY_DELAY = os.getenv("RETRY_DELAY", default=5) # Configurable for more conjested networks. 5 generally works.
```
The default needs to be provided as a string. If a user does not supply a value, and it tries to use the default which is already an integer, the `isdigit()` function is not available so it errors out.

```python
RETRY_DELAY = os.getenv("RETRY_DELAY", default="5")
```

NicJames2378 replied:
Good find. I've fixed this in my branch. I don't generally use dynamically typed languages, so some of the intricacies are lost on me.

brandan-schmitz commented on netbox_api.py:
```diff
@@ -147,37 +179,63 @@ def create_module_types(self, module_types):
+                            self.counter.update({'module_added': 1})
+                            self.handle.verbose_log(f'Module Type Created: {module_type_res.manufacturer.name} - {module_type_res.model} - {module_type_res.id}')
+                        except pynetbox.RequestError as exce:
+                            self.handle.log(f"Error '{exce.error}' creating module type: {curr_mt["manufacturer"]} {curr_mt["model"]} {curr_mt["part_number"]}")
```
Some things here need to be in single quotes not double:
```python
self.handle.log(f"Error '{exce.error}' creating module type: {curr_mt['manufacturer']} {curr_mt['model']} {curr_mt['part_number']}")
```

NicJames2378 replied:
Commit message didn't take for whatever reason. Fixed in commit 3ea22ac. Changed all [" to [' and all "] to ']

gurubert replied:
Either have outer ' and inner " or the other way around.
But you cannot have something like `f'Processing Source File: {curr_mt['src']}'`

It has to be `f'Processing Source File: {curr_mt["src"]}'` or `f"Processing Source File: {curr_mt['src']}"`

brandan-schmitz reviewed (changes requested):
I tried running your PR and I encountered a few issues. I outlined what I did to fix those here.

brandan-schmitz commented:
It also seems that some of the changes for retries upon connection errors is not working 100% as I have had instances where the Netbox server gets temporarily overloaded and returns a 502 connection error and the importer is not catching this and it causes it to crash.

```python
Traceback (most recent call last):
  File "/app/nb-dt-import.py", line 68, in <module>
    main()
  File "/app/nb-dt-import.py", line 30, in main
    netbox.create_device_types(device_types)
  File "/app/netbox_api.py", line 138, in create_device_types
    self.device_types.create_interfaces(device_type["interfaces"], dt.id)
  File "/app/netbox_api.py", line 281, in create_interfaces
    existing_interfaces = {str(item): item for item in self.netbox.dcim.interface_templates.filter(
  File "/app/netbox_api.py", line 281, in <dictcomp>
    existing_interfaces = {str(item): item for item in self.netbox.dcim.interface_templates.filter(
  File "/usr/local/lib/python3.9/site-packages/pynetbox/core/response.py", line 128, in __next__
    next(self.response), self.endpoint.api, self.endpoint
  File "/usr/local/lib/python3.9/site-packages/pynetbox/core/query.py", line 292, in get
    req = self._make_call(add_params=add_params)
  File "/usr/local/lib/python3.9/site-packages/pynetbox/core/query.py", line 259, in _make_call
    raise RequestError(req)
pynetbox.core.query.RequestError: The request failed with code 502 Bad Gateway but more specific details were not returned in json. Check the NetBox Logs or investigate this exception's error attribute.
```

NicJames2378 committed:
Fix RETRY_DELAY typing

The default value needs to be a string to prevent conversion errors, as we assume the user provided variable is a string (due to the way DotEnv handle variables)

NicJames2378 committed:
Update netbox_api.py

NicJames2378 commented:
I'm having trouble getting the 502 issue to reproduce on a fresh test environment, so that one may have to wait a week or so for me to get back to my dev environment. To assist me in reproducing, could you provide the NetBox logs from the time around the import? Additionally, can you confirm NetBox is still usable from the web both before and after performing an import? This is to verify that is a connectivity issue to the importer and not an issue in NetBox itself going offline.


As an alternative possibility, it looks like this may possibly be solved by amending netbox_api.py on line 164 to read as `except (http.client.RemoteDisconnected, requests.exceptions.ConnectionError, pynetbox.RequestError) as e:` instead. This will only help if the issue is caused by the connection dropping after the device type is created but _before_ the interfaces/ports/etc are. I didn't account for nor encounter this possibility initially, so it's possible I failed to catch a disconnection.

Let me know if you try the latter, and if it works I'll update the pull request with the new code.

bastianleicht reviewed (approved):
Worked on my end without Problems

DanSheps committed:
Merge branch 'develop' into develop

NicJames2378 committed:
move logging message to correct section

NicJames2378 committed:
fix importing

NicJames2378 committed:
add script to locate unused elevation images. Currently Pwsh, needs rewritten in Python.

NicJames2378 committed:
Merge branch 'develop' of https://github.com/NicJames2378/Device-Type-Library-Import into develop

NicJames2378 committed:
Merge branch 'import-fix' into develop
